cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1038,32 +1038,22 @@ public void batchNextRunTimes(FullHttpRequest request,
    */
   private List<BatchProgramSchedule> batchRunTimes(String namespace, Collection<? extends BatchProgram> programs,
                                                    boolean previous) throws Exception {
-    List<ProgramId> programIds = programs.stream()
-      .map(p -> new ProgramId(namespace, p.getAppId(), p.getProgramType(), p.getProgramId()))
+    List<ProgramReference> programReferences = programs.stream()
+      .map(p -> new ProgramReference(namespace, p.getAppId(), p.getProgramType(), p.getProgramId()))
       .collect(Collectors.toList());
-
-    Set<ApplicationId> appIds = programIds.stream().map(ProgramId::getParent).collect(Collectors.toSet());
-    Map<ApplicationId, ApplicationSpecification> appSpecs = store.getApplications(appIds);
+    Map<ProgramReference, ProgramId> programMap = store.getPrograms(programReferences);
 
     List<BatchProgramSchedule> result = new ArrayList<>();
-    for (ProgramId programId : programIds) {
-      ApplicationSpecification spec = appSpecs.get(programId.getParent());
-      if (spec == null) {
-        result.add(new BatchProgramSchedule(programId, HttpResponseStatus.NOT_FOUND.code(),
-                                            new NotFoundException(programId.getParent()).getMessage(), null));
-        continue;
-      }
-      try {
-        Store.ensureProgramExists(programId, spec);
+    for (ProgramReference programReference : programReferences) {
+      if (programMap.containsKey(programReference)) {
+        ProgramId programId = programMap.get(programReference);
         result.add(new BatchProgramSchedule(programId, HttpResponseStatus.OK.code(), null,
                                             getScheduledRunTimes(programId, previous)));
```

Comment:
both getScheduledRunTimes and BatchProgramSchedule should really be taking a ProgramReference, then we don't need to getPrograms().  but it is out of the scope of this PR.  I will continue the refactoring when have a chance